models/event.go

```diff
@@ -0,0 +1,81 @@
+package models
+
+import (
+	"fmt"
+	"time"
+)
+
+// A list of event types
+const (
+	EventFilterOutputCreated      = "FilterOutputCreated"
+	EventFilterOutputQueryStart   = "FilterOutputQueryStart"
+	EventFilterOutputQueryEnd     = "FilterOutputQueryEnd"
+	EventFilterOutputCSVGenStart  = "FilterOutputCSVGenStart"
+	EventFilterOutputCSVGenEnd    = "FilterOutputCSVGenEnd"
+	EventFilterOutputXLSXGenStart = "FilterOutputXLSXGenStart"
+	EventFilterOutputXLSXGenEnd   = "FilterOutputXLSXGenEnd"
+	EventFilterOutputCompleted    = "FilterOutputCompleted"
+)
+
+// Event captures the time certain stages of filter response were completed
+type Event struct {
+	Type string    `bson:"type,omitempty" json:"type"`
+	Time time.Time `bson:"time,omitempty" json:"time"`
+}
+
+func (filter *Filter) RemoveDuplicateEvents(currentFilter *Filter) {
+	if len(filter.Events) == 0 || len(currentFilter.Events) == 0 {
+		return
+	}
+
+	events := []*Event{}
+
+	for _, e := range filter.Events {
+		found := false
+		for _, ce := range currentFilter.Events {
+			//compare the values not the pointers
+			if *e == *ce {
+				found = true
+			}
+		}
```

Comment:
We could stop iterating the inner for loop once we found the event
```suggestion
		for _, ce := range currentFilter.Events {
			//compare the values not the pointers
			if *e == *ce {
				found = true
				break
			}
		}
```

Or slightly change the for loop if we prefer not to add `break` statements:

```suggestion
		for i :=0; !found && i < len(currentFilterEvents.Events; i++ {
		        ce := currentFilterEvents.Events[i]
			//compare the values not the pointers
			if *e == *ce {
				found = true
			}
		}
```